File: DailyChallenges/CPP/LeetCode_20.cpp
class Solution {
public:
    bool isValid(string s) {
        vector<char> v;
        for (auto ch:s){
            if ((ch == ')' && (!v.empty() && v.back() == '(')) ||
                (ch == ']' && (!v.empty() && v.back() == '[')) ||
                (ch == '}' && (!v.empty() && v.back() == '{'))){
                v.pop_back();
                continue;
            }
            if (ch == '(' || ch == '[' || ch == '{'){
                v.push_back(ch);
                continue;
            }
            return false;
        }
        return v.empty();
    }
};

File: DailyChallenges/Scala/LeetCode_20.scala
object Solution {
    def isValid(s: String): Boolean = {
        def check(stack: List[Char], str: List[Char]): Boolean = if (str.isEmpty) stack.isEmpty else {
            val c = str.head
            c match {
                case '{' => check(c::stack, str.tail)
                case '[' => check(c::stack, str.tail)
                case '(' => check(c::stack, str.tail)
                case ')' => if (stack.nonEmpty && stack.head == '(') check(stack.tail, str.tail) else false
                case '}' => if (stack.nonEmpty && stack.head == '{') check(stack.tail, str.tail) else false                       
                case ']' => if (stack.nonEmpty && stack.head == '[') check(stack.tail, str.tail) else false
                case _ => check(stack, str.tail)
            }
        }
        check(List.empty[Char], s.toList)
    }
}
